core/src/main/scala/kafka/zk/KafkaZkClient.scala

```diff
@@ -86,14 +89,64 @@ class KafkaZkClient private (zooKeeperClient: ZooKeeperClient, isSecure: Boolean
   }
 
   /**
-   * Registers a given broker in zookeeper as the controller.
+   * Registers a given broker in zookeeper as the controller and increments controller epoch.
+   * @return the (updated controller epoch, epoch zkVersion) tuple
    * @param controllerId the id of the broker that is to be registered as the controller.
-   * @param timestamp the timestamp of the controller election.
-   * @throws KeeperException if an error is returned by ZooKeeper.
+   * @throws ControllerMovedException if fail to create /controller or fail to increment controller epoch.
    */
-  def registerController(controllerId: Int, timestamp: Long): Unit = {
-    val path = ControllerZNode.path
-    checkedEphemeralCreate(path, ControllerZNode.encode(controllerId, timestamp))
+  def registerControllerAndIncrementControllerEpoch(controllerId: Int): (Int, Int) = {
+    val timestamp = time.milliseconds()
+
+    // Create /controller_epoch if not exists
+    maybeCreateControllerEpochZNode()
+
+    // Read /controller_epoch to get the current controller epoch and zkVersion
+    val (curEpoch, curEpochStat) = getControllerEpoch.getOrElse(
+      throw new IllegalStateException(s"${ControllerEpochZNode.path} does not exist while trying to create ${ControllerZNode.path}"))
+
+    // Create /controller and update /controller_epoch atomically
+    val newControllerEpoch = curEpoch + 1
+    val expectedControllerEpochZkVersion = curEpochStat.getVersion
+
+    debug(s"Try to create ${ControllerZNode.path} and increment controller epoch to $newControllerEpoch with expected controller epoch zkVersion $expectedControllerEpochZkVersion")
+    try {
+      val transaction = zooKeeperClient.createTransaction()
+      transaction.check(ControllerEpochZNode.path, expectedControllerEpochZkVersion)
+      transaction.create(ControllerZNode.path, ControllerZNode.encode(controllerId, timestamp),
+        acls(ControllerZNode.path).asJava, CreateMode.EPHEMERAL)
+      transaction.setData(ControllerEpochZNode.path, ControllerEpochZNode.encode(newControllerEpoch), expectedControllerEpochZkVersion)
+      val results = transaction.commit()
+      val setDataResult = results.get(2).asInstanceOf[SetDataResult]
+      (newControllerEpoch, setDataResult.getStat.getVersion)
+    } catch {
+      case _: NodeExistsException =>
+        val curControllerId = getControllerId.getOrElse(throw new ControllerMovedException(
+          s"The ephemeral node at ${ControllerZNode.path} went away while checking whether the controller election succeeds. " +
+          s"Aborting controller startup procedure"))
+        if (controllerId == curControllerId) {
+          val (epoch, stat)  = getControllerEpoch.getOrElse(throw new ControllerMovedException(
```

Comment:
Hmm, I am not sure this is safe. The controller path could have been deleted and grabbed by another broker in the window between line 123 and here. Then, we would have grabbed the wrong controller epoch.

Reply:
Ah. Ok. That's fine then.